src/charm.py

```diff
@@ -344,6 +348,41 @@ def _on_install(self, _event: InstallEvent) -> None:
         """
         state = self._setup_state()
 
+        if state.charm_config.openstack_clouds_yaml:
+            # Only build it in test mode since it may interfere with users systems.
+            if self.config.get("test-mode") == "insecure":
+                self.unit.status = MaintenanceStatus("Building Openstack image")
+                github = GithubClient(token=state.charm_config.token)
+                image = openstack_manager.build_image(
+                    arch=state.arch,
+                    cloud_config=state.charm_config.openstack_clouds_yaml,
+                    github_client=github,
+                    path=state.charm_config.path,
+                    proxies=state.proxy_config,
+                )
+                instance_config = openstack_manager.create_instance_config(
+                    unit_name=self.unit.name,
+                    openstack_image=image,
+                    path=state.charm_config.path,
+                    github_client=github,
+                )
+                self.unit.status = MaintenanceStatus("Creating Openstack test instance")
+                instance = openstack_manager.create_instance(
+                    cloud_config=state.charm_config.openstack_clouds_yaml,
+                    instance_config=instance_config,
+                    proxies=state.proxy_config,
+                    dockerhub_mirror=state.charm_config.dockerhub_mirror,
+                    ssh_debug_connections=state.ssh_debug_connections,
+                )
+                logger.info("OpenStack instance: %s", instance)
+            # Test out openstack integration and then go
+            # into BlockedStatus as it is not supported yet
+            self.unit.status = BlockedStatus(
```

Comment:
Could this call the shared `_block_on_openstack_config` function?

Reply:
Updated, thank you!